osu.Game/Graphics/UserInterface/ShearedSearchTextBox.cs

```diff
@@ -0,0 +1,192 @@
+                            {
+                                new Container
+                                {
+                                    Name = @"Search box container",
+                                    RelativeSizeAxes = Axes.Both,
+                                    CornerRadius = corner_radius,
+                                    Masking = true,
+                                    Children = new Drawable[]
+                                    {
+                                        new Box
+                                        {
+                                            Colour = colourProvider.Background4,
+                                            RelativeSizeAxes = Axes.Both
+                                        },
+                                        TextBox = new SearchTextBox
+                                        {
+                                            Shear = -shear,
+                                            Anchor = Anchor.CentreLeft,
+                                            Origin = Anchor.CentreLeft,
+                                            RelativeSizeAxes = Axes.X,
+                                            Padding = new MarginPadding
+                                            {
+                                                Horizontal = corner_radius
+                                            }
+                                        }
+                                    }
+                                },
+                                new Container
+                                {
+                                    Name = @"Icon container",
+                                    RelativeSizeAxes = Axes.Y,
+                                    Width = icon_container_width,
+                                    Origin = Anchor.CentreRight,
+                                    Anchor = Anchor.CentreRight,
+                                    Children = new Drawable[]
+                                    {
+                                        new SpriteIcon
+                                        {
+                                            Icon = FontAwesome.Solid.Search,
+                                            Origin = Anchor.Centre,
+                                            Anchor = Anchor.Centre,
+                                            Size = new Vector2(16),
+                                            Shear = -shear
+                                        }
+                                    }
+                                }
+                            }
+                        },
+                        ColumnDimensions = new[] { new Dimension(), new Dimension(GridSizeMode.AutoSize) }
+                    }
+                }
+            };
+        }
+
+        public override bool HandleNonPositionalInput => TextBox.HandleNonPositionalInput;
+
+        private class SearchTextBox : FocusedTextBox
```

Comment:
Why is this a local copy [of the existing `SearchTextBox`](https://github.com/ppy/osu/blob/master/osu.Game/Graphics/UserInterface/SearchTextBox.cs)? Surely it'd be easier to inherit from that and only change background colour as you see fit?

Reply:
existing `SearchTextBox` has an internal `SpriteIcon` to display the search icon. can inherit the existing one and remove the `SpriteIcon` if it's a better approach.

```diff
diff --git a/osu.Game/Graphics/UserInterface/ShearedSearchTextBox.cs b/osu.Game/Graphics/UserInterface/ShearedSearchTextBox.cs
index e3dc118faa..0d5a5e1b15 100644
--- a/osu.Game/Graphics/UserInterface/ShearedSearchTextBox.cs
+++ b/osu.Game/Graphics/UserInterface/ShearedSearchTextBox.cs
@@ -1,6 +1,7 @@
 // Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
 // See the LICENCE file in the repository root for full licence text.

+using System.Linq;
 using osu.Framework.Allocation;
 using osu.Framework.Bindables;
 using osu.Framework.Graphics;
@@ -8,22 +9,20 @@
 using osu.Framework.Graphics.Shapes;
 using osu.Framework.Graphics.Sprites;
 using osu.Framework.Graphics.UserInterface;
-using osu.Framework.Input;
-using osu.Framework.Input.Events;
 using osu.Game.Graphics.Sprites;
 using osu.Game.Overlays;
 using osuTK;
 using osuTK.Graphics;
-using osuTK.Input;

 namespace osu.Game.Graphics.UserInterface
 {
     public class ShearedSearchTextBox : CompositeDrawable, IHasCurrentValue<string>
     {
         private const float icon_container_width = 50;
         private const float corner_radius = 10;
         private const float height = 42;
         private readonly Vector2 shear = new Vector2(0.2f, 0);

         public FocusedTextBox TextBox;

@@ -70,7 +69,7 @@ private void load(OverlayColourProvider colourProvider)
                                             Colour = colourProvider.Background4,
                                             RelativeSizeAxes = Axes.Both
                                         },
-                                        TextBox = new SearchTextBox
+                                        TextBox = new NoIconSearchTextBox
                                         {
                                             Shear = -shear,
                                             Anchor = Anchor.CentreLeft,
@@ -78,7 +77,7 @@ private void load(OverlayColourProvider colourProvider)
                                             RelativeSizeAxes = Axes.X,
                                             Padding = new MarginPadding
                                             {
                                                 Horizontal = corner_radius
                                             }
                                         }
                                     }
@@ -112,10 +111,8 @@ private void load(OverlayColourProvider colourProvider)

         public override bool HandleNonPositionalInput => TextBox.HandleNonPositionalInput;

-        private class SearchTextBox : FocusedTextBox
+        private class NoIconSearchTextBox : SearchTextBox
         {
-            protected virtual bool AllowCommit => false;
-
             [BackgroundDependencyLoader]
             private void load(OverlayColourProvider colourProvider)
             {
@@ -123,6 +120,9 @@ private void load(OverlayColourProvider colourProvider)
                 BackgroundUnfocused = colourProvider.Background4;
                 Placeholder.Colour = Color4.White;
                 PlaceholderText = @"Search";
+
+                Remove(Children.OfType<SpriteIcon>().First());
+                TextFlow.Padding = new MarginPadding(0f);
             }

             protected override SpriteText CreatePlaceholder() => new OsuSpriteText
@@ -135,58 +135,6 @@ private void load(OverlayColourProvider colourProvider)
                 AutoSizeAxes = Axes.Both,
                 Child = new OsuSpriteText { Text = c.ToString(), Font = OsuFont.GetFont(size: 20, weight: FontWeight.SemiBold) },
             };
-
-            public override bool OnPressed(KeyBindingPressEvent<PlatformAction> e)
-            {
-                switch (e.Action)
-                {
-                    case PlatformAction.MoveBackwardLine:
-                    case PlatformAction.MoveForwardLine:
-                    // Shift+delete is handled via PlatformAction on macOS. this is not so useful in the context of a SearchTextBox
-                    // as we do not allow arrow key navigation in the first place (ie. the caret should always be at the end of text)
-                    // Avoid handling it here to allow other components to potentially consume the shortcut.
-                    case PlatformAction.DeleteForwardChar:
-                        return false;
-                }
-
-                return base.OnPressed(e);
-            }
-
-            protected override bool OnKeyDown(KeyDownEvent e)
-            {
-                if (!e.ControlPressed && !e.ShiftPressed)
-                {
-                    switch (e.Key)
-                    {
-                        case Key.Left:
-                        case Key.Right:
-                        case Key.Up:
-                        case Key.Down:
-                            return false;
-                    }
-                }
-
-                if (!AllowCommit)
-                {
-                    switch (e.Key)
-                    {
-                        case Key.KeypadEnter:
-                        case Key.Enter:
-                            return false;
-                    }
-                }
-
-                if (e.ShiftPressed)
-                {
-                    switch (e.Key)
-                    {
-                        case Key.Delete:
-                            return false;
-                    }
-                }
-
-                return base.OnKeyDown(e);
-            }
         }
     }

```

Reply:
Inheritance is definitely not the right direction. What I'd probably do is apply composition, i.e.:

* Split `SearchTextBox` into two classes, one being the textbox and textbox only, and the second containing that textbox and combining it with the icon (following a similar structure that you've used for the sheared one)
* Replace all existing usages of `SearchTextBox` with instances of the second class

Reply:
Also cc @jai-x because I recall you being interested in these search boxes before too...

Reply:
Agree with above.
It would be better to make a PR to modify `SearchTextBox` class such that it reatains the specific key behaviour but with no icon or visual customisation present. It can then be composed more easily in components like this sheared one or for the existing search text box design.
I had decided not to do that option as part the chat design implemntation since it would have required managing two textbox instances but it's more useful in the general case.